### PR TITLE
Add rollout upgrade integration test

### PR DESCRIPTION
**Pull request by bschimke95**
# Summary

Adds a test that creates a HA workload cluster and upgrades it from 1.29 Kubernetes to 1.30.

# Implementation details

* Non-HA clusters require in-place updates which are currently not supported
* We don't have snapd on the docker infra and cannot install different versions via `snap install`. That's why we build two different docker images and replace the infrastructure reference accordingly along with the upgrade. On other infra providers we would simply change the version and not the image.
* Right now, we inject a Kubernetes override into the docker build process as we don't have another minor release besides 1.30. Once 1.31 is out, we can bump the test versions, upgrade from 1.30 -> 1.31 and remove this hack in the build process.

# Notes
This tests require https://github.com/canonical/microcluster/pull/197 to land in microcluster and https://github.com/canonical/k8s-snap/pull/560 and https://github.com/canonical/k8s-snap/pull/562 to land in k8s-snap.

### CODE REVIEW

File: test/e2e/cluster_upgrade.go [hunk 1]
```diff
@@ -157,6 +157,7 @@ func ClusterUpgradeSpec(ctx context.Context, inputGetter func() ClusterUpgradeSp
 			ClusterProxy:                input.BootstrapClusterProxy,
 			Cluster:                     result.Cluster,
 			ControlPlane:                result.ControlPlane,
+			UpgradeMachineTemplate:      ptr.To(fmt.Sprintf("%s-control-plane-1.30", clusterName)),
```
Comment (addyess) [inline test/e2e/cluster_upgrade.go:160]: Does it feel odd to hard-code these version numbers into the test?  I suppose it's fine b/c it could be any version right?  like 1.xx

**Review by addyess — COMMENTED** [review]
Just comments.  I don't feel like I can give better feedback

File: test/e2e/cluster_upgrade.go (hunk 1 above)
Comment (bschimke95) [reply test/e2e/cluster_upgrade.go:160]: Yes, we don't test a particular version but rather if the upgrade mechanism of the provider is working as expected.

File: pkg/ck8s/workload_cluster.go [hunk 2]
```diff
@@ -281,8 +279,8 @@ func (w *Workload) doK8sdRequest(ctx context.Context, method, endpoint string, r
 	if responseBody.Error != "" {
 		return fmt.Errorf("k8sd request failed: %s", responseBody.Error)
 	}
-	if responseBody.Metadata == nil {
-		// Nothing to decode
+	if responseBody.Metadata == nil || response == nil {
```
Comment (bschimke95) [inline pkg/ck8s/workload_cluster.go:282]: drive-by fix: This would not cause issues but pollutes the logs with an error.

File: test/e2e/cluster_upgrade_test.go [hunk 3]
```diff
@@ -25,12 +25,11 @@ import (
...
 
 var _ = Describe("Workload cluster upgrade [CK8s-Upgrade]", func() {
-	BeforeEach(func() {
-		// TODO(bschimke): Remove once we find a way to run e2e tests with other infrastructure providers that support snap.
-		Skip("Skipping the upgrade tests as snap does not work on CAPD.")
-	})
-
 	Context("Upgrading a cluster with 1 control plane", func() {
+		It("Non-HA upgrades require in-place upgrades which are not supported yet.",
+			// TODO(ben): Enable this test once we have support for in-place upgrades.
```
Comment (bschimke95) [inline test/e2e/cluster_upgrade_test.go:30]: @berkayoz This is the test we need to enable again once the in-place upgrades are there.

File: templates/aws/ccm.yaml
Comment (neoaggelos) [inline templates/aws/ccm.yaml:1]: this in unrelated for this PR

File: templates/docker/Dockerfile [hunk 4]
```diff
@@ -36,15 +36,21 @@ FROM $BUILD_BASE AS builder
 ARG REPO=https://github.com/canonical/k8s-snap
 ARG BRANCH=main
 
+## Override the Kubernetes version from the branch.
+## Note(ben): We only have 1.30-release branches for k8s-snap right now.
+## For the rollout upgrades, we need to have a different minor version.
+## This is a temporary solution until we have a 1.31 release branch.
+ARG KUBERNETES_VERSION_OVERRIDE=""
```
Comment (neoaggelos) [inline templates/docker/Dockerfile:43]: `ARG KUBERNETES_VERSION` should suffice, and no need for the extra comments. It's OK to have a build argument to specify the exact version if we want

File: templates/docker/Dockerfile [hunk 5]
```diff
@@ -86,6 +92,12 @@ RUN /src/k8s-snap/build-scripts/build-component.sh helm
 
 ## kubernetes build
 FROM builder AS build-kubernetes
+ENV KUBERNETES_VERSION_OVERRIDE=${KUBERNETES_VERSION_OVERRIDE}
+RUN if [ -n "$KUBERNETES_VERSION_OVERRIDE" ]; then \
+      echo "Overwriting Kubernetes version with $KUBERNETES_VERSION_OVERRIDE"; \
+      echo "$KUBERNETES_VERSION_OVERRIDE" > /src/k8s-snap/build-scripts/components/kubernetes/version; \
+      cat /src/k8s-snap/build-scripts/components/kubernetes/version; \
```
Comment (neoaggelos) [inline templates/docker/Dockerfile:99]: this can be just:
```
[ -n "$KUBERNETES_VERSION" ] && echo "$KUBERNETES_VERSION" > /src/k8s-snap/build-scripts/components/kubernetes/version
```

File: test/e2e/data/infrastructure-docker/cluster-template-upgrades.yaml [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: DockerMachineTemplate
+metadata:
+  name: ${CLUSTER_NAME}-control-plane-1.30
+  namespace: ${NAMESPACE}
+spec:
+  template:
+    spec:
+      customImage: k8s-snap:dev-1.30
```
Comment (neoaggelos) [inline test/e2e/data/infrastructure-docker/cluster-template-upgrades.yaml:73]: why do we deploy a second machine template? we should simply apply the updated version of the `DockerMachineTemplate` resource

File: test/e2e/helpers.go [hunk 7]
```diff
@@ -577,6 +578,17 @@ func UpgradeControlPlaneAndWaitForUpgrade(ctx context.Context, input UpgradeCont
...
+	newInfrastructureRef := corev1.ObjectReference{
+		APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1", // Adjust based on your infrastructure API version
+		Kind:       "DockerMachineTemplate",
+		Name:       fmt.Sprintf("%s-control-plane-1.30", input.Cluster.Name),
+		Namespace:  input.ControlPlane.Spec.MachineTemplate.InfrastructureRef.Namespace,
+	}
+
+	// Update the infrastructureRef
+	input.ControlPlane.Spec.MachineTemplate.InfrastructureRef = newInfrastructureRef
+
```
Comment (neoaggelos) [inline test/e2e/helpers.go:591]: we should instead just update the machine template with the new image

File: test/e2e/config/ck8s-docker.yaml [hunk 8]
```diff
@@ -84,8 +85,8 @@ providers:
 
 variables:
   KUBERNETES_VERSION_MANAGEMENT: "v1.28.0"
-  KUBERNETES_VERSION: "v1.30.0"
-  KUBERNETES_VERSION_UPGRADE_TO: "v1.30.1"
+  KUBERNETES_VERSION: "v1.29.6"
+  KUBERNETES_VERSION_UPGRADE_TO: "v1.30.2"
```
[the remark this answers is not in the document]
Comment (bschimke95) [reply test/e2e/config/ck8s-docker.yaml:89]: This actually does not make any difference.
We only compare for the minor version and build the relevant images in the e2e.yaml Github workflow rather then using this value here.

File: .github/workflows/e2e.yaml [hunk 9]
```diff
@@ -11,6 +11,14 @@ jobs:
...
+        name: Login to GitHub Container Registry
+        uses: docker/login-action@v3
+        with:
+          # We run into rate limiting issues if we don't authenticate
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
       - name: Check out repo
         uses: actions/checkout@v4
       - name: Install requirements
```
Comment (neoaggelos) [inline .github/workflows/e2e.yaml:24]: Just a note: make sure to avoid too high coupling between the workflow itself and the setting up of the CI (e.g. runners). It works now, but the danger would be that it becomes very brittle and one of those things that people are afraid to touch.

File: .github/workflows/e2e.yaml [hunk 10]
```diff
@@ -25,26 +33,30 @@ jobs:
       - name: Build k8s-snap image
         run: |
           cd templates/docker
-          sudo docker build . -t k8s-snap:dev
+          sudo docker build . -t k8s-snap:dev-old --build-arg BRANCH=main --build-arg KUBERNETES_VERSION=v1.29.6
+          sudo docker build . -t k8s-snap:dev-new --build-arg BRANCH=main
```
Comment (neoaggelos) [inline .github/workflows/e2e.yaml:37]: should this be in a wrapper script? where you build all necessary images? this version will need to be updated eventually, it would rather not be a hardcoded value on a github action

File: test/e2e/config/ck8s-docker.yaml [hunk 11]
```diff
@@ -84,8 +85,8 @@ providers:
 
 variables:
   KUBERNETES_VERSION_MANAGEMENT: "v1.28.0"
-  KUBERNETES_VERSION: "v1.30.0"
-  KUBERNETES_VERSION_UPGRADE_TO: "v1.30.1"
+  KUBERNETES_VERSION: "v1.29.6"
+  KUBERNETES_VERSION_UPGRADE_TO: "v1.30.3"
   IP_FAMILY: "IPv4"
```
Comment (neoaggelos) [inline test/e2e/config/ck8s-docker.yaml:90]: especially when we already repeat the configs here. i can already see confusion about "i updated this value, why does it not work as i expect?"

**Review by neoaggelos — COMMENTED** [review]
Overall LGTM, left some comments around structure

File: .github/workflows/e2e.yaml (hunk 9 above)
Comment (bschimke95) [reply .github/workflows/e2e.yaml:24]: You mean the Github login should not be a separate step? Where would you put this instead?

File: test/e2e/config/ck8s-docker.yaml [hunk 12]
```diff
@@ -51,9 +51,10 @@ providers:
           - old: "imagePullPolicy: Always"
             new: "imagePullPolicy: IfNotPresent"
     files:
-      - sourcePath: "../data/infrastructure-docker/cluster-template.yaml"
       - sourcePath: "../data/infrastructure-docker/cluster-template-kcp-remediation.yaml"
       - sourcePath: "../data/infrastructure-docker/cluster-template-md-remediation.yaml"
+      - sourcePath: "../data/infrastructure-docker/cluster-template-upgrades.yaml"
+      - sourcePath: "../data/infrastructure-docker/cluster-template.yaml"
```
Comment (HomayoonAlimohammadi) [inline test/e2e/config/ck8s-docker.yaml:57]: Just being curious, does the ordering matter here?

File: test/e2e/data/infrastructure-docker/cluster-template-md-remediation.yaml [hunk 13]
```diff
@@ -57,7 +57,7 @@ metadata:
 spec:
   template:
     spec:
-      customImage: k8s-snap:dev
+      customImage: k8s-snap:dev-old
```
Comment (HomayoonAlimohammadi) [inline test/e2e/data/infrastructure-docker/cluster-template-md-remediation.yaml:60]: again, just being curious, is there any particular reason that we used `dev-old` here and `dev-new` in the other non-upgrade templates (e.g. `cluster-template-kcp-remediation.yaml`)?

File: test/e2e/data/infrastructure-docker/cluster-template-upgrades.yaml [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+      kind: DockerMachineTemplate
+      name: ${CLUSTER_NAME}-control-plane-old
+  spec:
+    airGapped: true
+    controlPlane:
+      extraKubeAPIServerArgs:
+        --anonymous-auth: "true"
+  replicas: ${CONTROL_PLANE_MACHINE_COUNT}
+  version: ${KUBERNETES_VERSION}
+# Initial template for the machine deployment
```
Comment (HomayoonAlimohammadi) [inline test/e2e/data/infrastructure-docker/cluster-template-upgrades.yaml:51]: maybe we can move this to the next line (after `---`)? I think it's more readable that way.

File: test/e2e/data/infrastructure-docker/cluster-template-upgrades.yaml [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: DockerMachineTemplate
+metadata:
+  name: ${CLUSTER_NAME}-control-plane-old
+  namespace: ${NAMESPACE}
+spec:
+  template:
+    spec:
+      customImage: k8s-snap:dev-old
+# After upgrade template for the machine deployment
```
Comment (HomayoonAlimohammadi) [inline test/e2e/data/infrastructure-docker/cluster-template-upgrades.yaml:62]: this as well

File: test/e2e/data/infrastructure-docker/cluster-template-upgrades.yaml [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+        configRef:
+          apiVersion: bootstrap.cluster.x-k8s.io/v1beta2
+          kind: CK8sConfigTemplate
+          name: ${CLUSTER_NAME}-md-0
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+        kind: DockerMachineTemplate
+        name: ${CLUSTER_NAME}-md-old-0
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
```
Comment (HomayoonAlimohammadi) [inline test/e2e/data/infrastructure-docker/cluster-template-upgrades.yaml:107]: do we need to add any comments for worker nodes as well?

File: test/e2e/helpers.go [hunk 17]
```diff
@@ -607,14 +661,17 @@ func WaitForNodesReady(ctx context.Context, input WaitForNodesReadyInput) {
 		}
 		nodeReadyCount := 0
 		for _, node := range nodeList.Items {
+			fmt.Fprintf(GinkgoWriter, "versions: %s %s\n", semver.MajorMinor(node.Status.NodeInfo.KubeletVersion), semver.MajorMinor(input.KubernetesVersion))
```
Comment (HomayoonAlimohammadi) [inline test/e2e/helpers.go:664]: maybe we can add more context to this line, something like:
```suggestion
			fmt.Fprintf(GinkgoWriter, "KubeletVersion: %s, KubernetesVersion: %s\n", semver.MajorMinor(node.Status.NodeInfo.KubeletVersion), semver.MajorMinor(input.KubernetesVersion))
```

**Review by HomayoonAlimohammadi — APPROVED** [review]
LGTM, Thanks a lot @bschimke95! Just some minor comments and questions.

File: test/e2e/config/ck8s-docker.yaml (hunk 12 above)
Comment (bschimke95) [reply test/e2e/config/ck8s-docker.yaml:57]: I don't think so, these are just the specifications of template files.

File: test/e2e/data/infrastructure-docker/cluster-template-md-remediation.yaml (hunk 13 above)
Comment (bschimke95) [reply test/e2e/data/infrastructure-docker/cluster-template-md-remediation.yaml:60]: Not really, let's use `dev-old` everywhere.
